src/PreallocationTools.jl

```diff
@@ -8,7 +8,7 @@ struct DiffCache{T<:AbstractArray, S<:AbstractArray}
 end
 
 function DiffCache(u::AbstractArray{T}, siz, ::Type{Val{chunk_size}}) where {T, chunk_size}
-    x = ArrayInterface.restructure(u,zeros(ForwardDiff.Dual{nothing,T,chunk_size}, siz...))
+    x = zeros(T,(chunk_size+1)*prod(siz))  

```

Comment:
Made a new commit reflecting this suggestion and included (a lot) more tests. It should work as intended with GPU arrays now (i.e., dual_du is always a GPU array). The performance in the toy examples that I tested was indistinguishable from master.

One thing that does **not** work anymore is generating a cache with abstract typed elements (because sizeof() is not defined for an abstract type... of course).

Reply:
Oh, and I am also wondering whether the `DiffCache` constructor signature could now be simplified from:
`function DiffCache(u::AbstractArray{T}, siz, ::Type{Val{chunk_size}}) where {T, chunk_size}`
to
`function DiffCache(u::AbstractArray{T}, siz, chunk_size::Int) where {T}`

because the element type of the cache now does not depend on the chunk_size anymore?

Reply:
Yes, I guess it can.